detectors/resources-support/src/main/java/com/google/cloud/opentelemetry/detectors/GCPPlatformDetector.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.opentelemetry.detectors;
+
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+public class GCPPlatformDetector {
+  public static final GCPPlatformDetector DEFAULT_INSTANCE = new GCPPlatformDetector();
+
+  private final GCPMetadataConfig metadataConfig;
+  private final EnvironmentVariables environmentVariables;
+
+  // for testing only
+  GCPPlatformDetector(GCPMetadataConfig metadataConfig, EnvironmentVariables environmentVariables) {
+    this.metadataConfig = metadataConfig;
+    this.environmentVariables = environmentVariables;
+  }
+
+  private GCPPlatformDetector() {
+    this.metadataConfig = GCPMetadataConfig.DEFAULT_INSTANCE;
+    this.environmentVariables = EnvironmentVariables.DEFAULT_INSTANCE;
+  }
+
+  // Detects the GCP platform on which the application is running
+  public DetectedPlatform detectPlatform() {
+    if (!isRunningOnGcp()) {
+      return generateDetectedPlatform(SupportedPlatform.UNKNOWN_PLATFORM);
+    }
+    Function<EnvironmentVariables, Optional<SupportedPlatform>> detectGKE =
```

Comment:
I'm not sure how this is more readable than an if/else statement.  Is there a reason to construct all these thunks?

Reply:
In this case, the detection functions ended up being one-liners, so if-else seems like a viable replacement, but in case the platform detection was a more involved process, I feel this approach would be more readable one. 

IMO, it also streamlines the process of running detection one-by-one on each platform.

Reply:
I agree with the additional pressure on GC point. Personally, I do not prefer if-else-if ladders but as you mentioned detection logic is unlikely to be changed - so I will update the code to remove the additional Functions.